parser.go

```
@@ -0,0 +1,191 @@
+var ErrInvalidSectionName error
+
+// ErrCouldNotWriteToFile happens when file cannot be written to
+var ErrCouldNotWriteToFile error
+
+// LoadFromFile loads ini file
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromFile(fileName string) error {
+	var iniLines []string
+	var input io.Reader
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		ErrCouldNotOpen = fmt.Errorf("cannot open file: %s", fileName)
+		return ErrCouldNotOpen
+	}
+	defer file.Close()
+
+	input = file
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		iniLines = append(iniLines, scanner.Text())
+	}
+	return parsedMap.parserLogic(iniLines)
+}
+
+// LoadFromString loads ini script from a string
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromString(str string) error {
+	iniLines := strings.Split(str, "\n")
+	return parsedMap.parserLogic(iniLines)
+}
+
+func (parsedMap *Parser) parserLogic(iniLines []string) error {
+	var section, key, value string
+
+	for _, line := range iniLines {
+		if len(line) == 0 {
+			continue
+		} else if line[0] == '[' {
+			closingParaFound := false
+			for j, ch := range line {
+				if ch == ']' {
+					section = line[1:j]
+					section = strings.Trim(section, " ")
+					closingParaFound = true
+					if parsedMap.dictionary[section] == nil {
+						parsedMap.dictionary[section] = make(map[string]string)
+					}
+				}
+			}
+			if !closingParaFound {
+				ErrSectionNameMissingClosure = fmt.Errorf("section [%s] is missing closure paranthesis ']'", line[1:])
+				return ErrSectionNameMissingClosure
+			}
+			if parsedMap.dictionary == nil {
+				parsedMap.dictionary = make(map[string]map[string]string)
+			}
+
+		} else if line[0] == ']' {
+			ErrWrongParanthesisOrder = fmt.Errorf("wrong section paranthesis order, the section provided, %s, cannot start by ]", line)
+			return ErrWrongParanthesisOrder
```

you shouldn't add a special case of line starting with ], it's simply a format error the else at the end of if... else if should handle this

---

> you shouldn't add a special case of line starting with ], it's simply a format error the else at the end of if... else if should handle this

I kinda wanted to cover edgy corner cases, since issues stated that keys can start with literals, this would be wrong to tell user that your key is missing value, when actually it's a section missing the opening paranthesis,  
KEY //error no =
KEY] --> error [KEY]

---

we have few specific cases for a valid line here

1- a section identifier (string that matches format "[section]")
2- a key value pair (key=value) -> this should be within a section to be valid, otherwise it's a formatting error
3- an empty line
4- a comment line

only one of this is considered a valid format, other than that you'd need to return an error invalid format

---

trying to handle every corner case would make you do alot of work and end up missing some corner cases